Civi/Token/TokenCompatSubscriber.php

```diff
@@ -148,7 +148,73 @@ public function onRender(TokenRenderEvent $e) {
+        if (!empty($e->context['contact'])) {
+          $e->string = \CRM_Utils_Token::replaceContactTokens($e->string, $e->context['contact'], TRUE, $greetingTokens, TRUE, $useSmarty);
+        }
+
+        \CRM_Utils_Token::removeNullContactTokens($e->string, $e->context['contact'], $greetingTokens);
+        // check if there are any unevaluated tokens
+        $greetingTokens = \CRM_Utils_Token::getTokens($e->string);
+
+        // $greetingTokens not empty, means there are few tokens which are not
+        // evaluated, like custom data etc
+        // so retrieve it from database
+        if (!empty($greetingTokens) && array_key_exists('contact', $greetingTokens)) {
+          $greetingsReturnProperties = array_flip(\CRM_Utils_Array::value('contact', $greetingTokens));
+          $greetingsReturnProperties = array_fill_keys(array_keys($greetingsReturnProperties), 1);
+          $contactParams = ['contact_id' => $e->context['contact']['contact_id']];
+
+          $greetingDetails = \CRM_Utils_Token::getTokenDetails($contactParams,
+            $greetingsReturnProperties,
+            FALSE, FALSE, NULL,
+            $greetingTokens,
+            NULL
+          );
+
+          // again replace tokens
+          $e->string = \CRM_Utils_Token::replaceContactTokens($e->string,
+            $greetingDetails,
+            TRUE,
+            $greetingTokens,
+            TRUE,
+            $useSmarty
+          );
+        }
+
+        // check if there are still any unevaluated tokens
+        $remainingTokens = \CRM_Utils_Token::getTokens($e->string);
+
+        // $greetingTokens not empty, there are customized or hook tokens to replace
+        if (!empty($remainingTokens)) {
+          // Fill the return properties array
+          $greetingTokens = $remainingTokens;
+          reset($greetingTokens);
+          $greetingsReturnProperties = [];
+          foreach ($greetingTokens as $value) {
+            $props = array_flip($value);
+            $props = array_fill_keys(array_keys($props), 1);
+            $greetingsReturnProperties = $greetingsReturnProperties + $props;
+          }
+          $contactParams = ['contact_id' => $e->context['contact']['contact_id']];
+          $greetingDetails = \CRM_Utils_Token::getTokenDetails($contactParams,
+            $greetingsReturnProperties,
+            FALSE, FALSE, NULL,
+            $greetingTokens,
+            NULL
+          );
+          // Prepare variables for calling replaceHookTokens
+          $categories = array_keys($greetingTokens);
+          [$contact] = $greetingDetails;
+          // Replace tokens defined in Hooks.
+          $e->string = \CRM_Utils_Token::replaceHookTokens($e->string, $contact[$e->context['contact']['contact_id']], $categories);
+        }
+      }
+
       $e->string = \CRM_Utils_Token::replaceContactTokens($e->string, $e->context['contact'], $isHtml, $e->message['tokens'], TRUE, $useSmarty);
```

Comment:
@totten this line here ^^ is the one #21414 removes - the rest is an 'expansion' of the line above it